File: elibrary.py
#! /usr/bin/python3

import sys
import re
import os.path
import atexit
import json
import random
from getpass import getpass
from typing import Union, Dict

import requests
from bs4 import BeautifulSoup


PROXY_LIST_URL = 'https://premproxy.com/ru/proxy-by-country/list-ip-port/%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D1%8F-01.htm'
AUTH_URL = 'https://elibrary.ru/start_session.asp'
AUTHOR_LIST_URL = 'https://elibrary.ru/authors.asp'
AUTHOR_ITEMS_URL = 'https://elibrary.ru/author_items.asp'
AUTHOR_REFS_URL = 'https://elibrary.ru/author_refs.asp'


class Config:
    def __init__(self):
        self._cookies = None
        self.proxies = None
        self._session = requests.Session()

    @property
    def cookies(self):
        if self._cookies is None:
            self.load_cookies()

        return self._cookies

    @cookies.setter
    def cookies(self, value: dict):
        self._cookies = value

    @property
    def session(self):
        return self._session

    def save_cookies(self):
        with open('cookies.json', 'w') as file:
            json.dump(self._cookies, file)

    def load_cookies(self):
        if os.path.exists('cookies.json'):
            with open('cookies.json', 'r') as file:
                self._cookies = json.load(file)
        else:
            self._cookies = {}


CONFIG = Config()


def get_random_proxy():
    response = requests.get(PROXY_LIST_URL)
    soup = BeautifulSoup(response.content, 'lxml')

    exec(soup.select_one('head script').contents[0])

    proxy_list = []

    for row in soup.select('#ipportlist li'):
        ip = row.text

        port_insertion_js_script = row.select_one('script').contents[0]
        port_insertion_sequence = re.search(r'^document\.write\(":"\+(.*)\)$', port_insertion_js_script).groups()[0]

        local_variables = locals()

        port = ''.join([str(local_variables[digit]) for digit in port_insertion_sequence.split('+')])

        proxy_list.append(f'{ip}:{port}')

    random_proxy = random.choice(proxy_list)

    return random_proxy


def proxy_request(method: str,
                  url: str,
                  headers: dict = None,
                  data: Union[dict, str, bytes, bytearray] = None) -> Union[requests.Response, None]:
    try:
        response = CONFIG.session.request(
            method,
            url,
            headers=headers,
            data=data,
            cookies=CONFIG.cookies,
        )
    except requests.exceptions.ConnectionError:
        print('Ошибка подключения к прокси-серверу')
        sys.exit()

    if response.status_code == 500 or response.url.endswith('page_error.asp'):
        print('Ошибка сервера, попробуйте повторить вход')
        CONFIG.cookies = {}
        authenticate(**prompt_login())
        return proxy_request(method, url, headers=headers, data=data)
    elif response.url.endswith('ip_blocked.asp'):
        print('Данный IP заблокирован на сервисе elibrary.ru')
        sys.exit()

    return response


def authenticate(login: str, password: str):
    response = proxy_request(
        'POST',
        AUTH_URL,
        data={'login': login, 'password': password},
    )

    CONFIG.cookies = response.cookies.get_dict()


def prompt_login() -> Dict[str, str]:
    login = input('Введите имя пользователя: ')
    password = getpass(f'Пароль для {login}: ')

    return {'login': login, 'password': password}


def resolve_author(lastname: str) -> Union[Dict[str, str], None]:
    authors_list = []

    authors_page = proxy_request(
        'POST',
        AUTHOR_LIST_URL,
        headers={
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36',
            'content-type': 'application/x-www-form-urlencoded'
        },
        data={
            'surname': lastname,
            'orgname': 'Санкт-Петербургский государственный университет телекоммуникаций им. проф. М.А. Бонч-Бруевича',
            'orgid': '1193',
            'metrics': '1',
            'order': '0',
            'sortorder': '0',
        },
    )

    authors_soup = BeautifulSoup(authors_page.content, 'lxml')

    authors_row_list = authors_soup.select('#restab tr[valign="top"]')

    if len(authors_row_list) == 0:
        return None

    for author_row in authors_row_list:
        author_name = author_row.select_one('td.midtext[align="left"] > font > b').text.strip()

        while True:
            is_requested_author = input(author_name + '? [Y/n]: ').lower().strip()

            if is_requested_author in ('y', ''):
                author_id = author_row.get('id', '').lstrip('a')

                return {
                    'id': author_id,
                    'name': author_name,
                }
            elif is_requested_author == 'n':
                break


def get_author():
    lastname = input('Введите фамилию автора: ')

    return resolve_author(lastname)


def get_pages_count(url, author):
    page = proxy_request(
        'GET',
        url,
        data={
            'authorid': author['id'],
        },
    )

    soup = BeautifulSoup(page.content, 'lxml')

    try:
        page_count_expression = soup.select('#pages td')[-1].select_one('a').get('href', '')
    except (IndexError, AttributeError, TypeError):
        return 1

    page_count = int(re.search(r'^javascript:goto_page\(([0-9]+)\)$', page_count_expression).groups()[0])

    return page_count


def get_author_items(author):
    page_count = get_pages_count(AUTHOR_ITEMS_URL, author)

    author_items = []

    for page_number in range(1, page_count + 1):
        print(f'Загрузка статей автора (страница {page_number} из {page_count})...')

        items_page = proxy_request(
            'POST',
            AUTHOR_ITEMS_URL,
            data={
                'authorid': author['id'],
                'pagenum': page_number,
            },
        )

        items_soup = BeautifulSoup(items_page.content, 'lxml')

        items_row_list = items_soup.select('#restab tr[valign="middle"]')

        for row in items_row_list:
            author_items.append({
                'id': row.get('id', '').lstrip('arw'),
                'title': row.select_one('td[align="left"] a').text.strip(),
                'authors': row.select_one('td[align="left"] i').text.strip().split(', '),
                'description': row.select('td[align="left"] font')[-1].text.strip(),
            })

    return author_items


def get_author_refs(author):
    page_count = get_pages_count(AUTHOR_REFS_URL, author)

    author_refs = []

    for page_number in range(1, page_count + 1):
        print(f'Загрузка ссылок на статьи автора (страница {page_number} из {page_count})...')

        refs_page = proxy_request(
            'POST',
            AUTHOR_REFS_URL,
            data={
                'authorid': author['id'],
                'pagenum': page_number,
            },
        )

        refs_soup = BeautifulSoup(refs_page.content, 'lxml')

        refs_row_list = refs_soup.select('#restab tr[valign="middle"]')

        for row in refs_row_list:
            author_refs.append({
                'id': row.get('id', '').lstrip('arw'),
                'count_number': row.select_one('td[align="center"] b').text.strip(),
                'source': row.select_one('td[align="left"] > font').text.strip(),
                'cite_item': row.select_one('td[align="left"] table .menug').text.strip(),
                'malformed': row.select_one('td[align="left"] > a') is None,
            })

    return author_refs


if __name__ == '__main__':
    atexit.register(CONFIG.save_cookies)

    CONFIG.session.proxies['http'] = CONFIG.session.proxies['https'] = 'socks5h://localhost:9050'

    # while True:
    #     use_proxy = input('Использовать прокси? [Y/n]: ').lower().strip()
    #
    #     if use_proxy in ('y', ''):
    #         CONFIG.proxies = {'https': input('Введите адрес прокси в формате "<протокол>://<адрес>[:<порт>]": ')}
    #
    #         print(f"Текущий адрес прокси: {CONFIG.proxies['https']}")
    #
    #         break
    #     elif use_proxy == 'n':
    #         break

    if 'SCookieID' not in CONFIG.cookies and 'SUserID' not in CONFIG.cookies:
        authenticate(**prompt_login())

    author = get_author()

    if author is None:
        print('Автора с таким именем не найдено')
        sys.exit()

    items = get_author_items(author)
    refs = get_author_refs(author)

    got_malformed_refs = False

    for ref in refs:
        if ref['malformed']:
            got_malformed_refs = True

            print('\nНайдена неверная ссылка на статью:')
            print('Порядковый номер ссылки в списке ссылок: ' + ref['count_number'])
            print('Название цитируемой статьи:\n' + ref['source'])
            print('Название ссылающейся статьи:\n' + ref['cite_item'])

    if not got_malformed_refs:
        print('Неверных ссылок не найдено')

File: elibrary.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const readline = require('readline');
const getpass = require('getpass');


const PROXY_LIST_URL = 'https://premproxy.com/ru/proxy-by-country/list-ip-port/%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D1%8F-01.htm';
const AUTHOR_LIST_URL = 'https://elibrary.ru/authors.asp';
const AUTHOR_ITEMS_URL = 'https://elibrary.ru/author_items.asp';
const AUTHOR_REFS_URL = 'https://elibrary.ru/author_refs.asp';


const create_read_stream = () => readline.createInterface({
    input: process.stdin,
    output: process.stdout
});


const get_random_proxy = async (browser) => {
    const page = await browser.newPage();
    await page.goto(PROXY_LIST_URL);

    const proxy_list = [];

    await Promise.all([
        page.waitForNavigation(),
        page.select('#xf5', '2')
    ]);

    const address_list = await page.evaluate(() => {
        const rows = document.querySelectorAll('.spy1xx[onmouseover] .spy1x[onmouseover]');
        const list = [];

        for (let row of rows) {
            list.push(row.querySelector('td').innerText);
        }

        return list;
    });

    await page.close();

    return address_list[Math.round(Math.random() * (address_list.length - 1))];
};


const authenticate = async (page, login, password) => {
    await page.goto('https://elibrary.ru/defaultx.asp');

    await page.focus('#login');
    await page.keyboard.type(login);

    await page.focus('#password');
    await page.keyboard.type(password);

    await page.click('input[name="knowme"]');

    const [response] = await Promise.all([
        page.waitForNavigation(),
        page.click('#win_login .butred')
    ]);

    if (response.status() === 500) {
        throw new Error('Данный аккаунт или IP-адрес заблокирован на сервисе elibrary.ru');
    }

    fs.writeFileSync('cookies.json', JSON.stringify(await page.cookies()));
};


const prompt_login = async () => {
    const stream = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return await new Promise(resolve => {
        stream.question('Введите имя пользователя: ', login => {
            stream.question(`Пароль для ${login}: `, password => {
                resolve({login, password});
            })
        });
    });
};


const resolve_author = async (page, lastname) => {
    await page.goto(AUTHOR_LIST_URL);

    await page.focus('#surname');
    await page.keyboard.type(lastname);
    await page.evaluate(() => {
        document.getElementById('orgname').value = 'Санкт-Петербургский государственный университет телекоммуникаций им. проф. М.А. Бонч-Бруевича';
    });

    await Promise.all([
        page.waitForNavigation(),
        page.click('#show_param .butred')
    ]);

    const author_list = await page.evaluate(() => {
        const author_rows = [];

        const author_row_list = document.querySelectorAll('#restab tr[valign="top"]');

        if (author_row_list.length === 0) {
            return null;
        }

        for (let row of author_row_list) {
            author_rows.push({
                id: row.id.slice(1),
                name: row.querySelector('td.midtext[align="left"] > font').textContent
            });
        }

        return author_rows;
    });

    if (author_list === null) {
        return null;
    }

    for (let author of author_list) {
        while (true) {
            const readStream = create_read_stream();

            const is_requested_author = await new Promise(resolve => {
                readStream.question(`${author.name}? [Y/n]: `, answer => resolve(answer.toLowerCase()))
            });

            if (['y', ''].includes(is_requested_author)) {
                return author;
            } else if (is_requested_author === 'n') {
                break
            }
        }
    }

    return null;
};


const get_author = async (page) => {
    const lastname = await new Promise(resolve => {
        create_read_stream().question('Введите фамилию автора: ', lastname => resolve(lastname));
    });

    return await resolve_author(page, lastname);
};


const get_author_items = async (page, author) => {
    const author_items = [];

    let page_number = 1;

    while (true) {
        await page.goto(`${AUTHOR_ITEMS_URL}?authorid=${author.id}&pagenum=${page_number}`);

        const has_next_page = await page.evaluate(() => {
            const pagination_elements = document.querySelectorAll('#pages td.mouse-hovergr');

            if (pagination_elements.length === 0) {
                return false;
            }

            return pagination_elements[pagination_elements.length - 1].querySelector('font') === null;
        });

        const node_console = console;

        const page_items = await page.evaluate(() => {
            const item_row_list = document.querySelectorAll('#restab tr[valign="middle"]');
            const item_list = [];

            if (item_row_list.length === 0) {
                return null;
            }

            node_console.log(`Загрузка статей автора (страница ${page_number})...`);

            for (let row of item_row_list) {
                const description_elements = row.querySelectorAll('td[align="left"] font');

                item_list.push({
                    id: row.id.slice(3),
                    title: row.querySelector('td[align="left"] a').textContent.trim(),
                    authors: row.querySelector('td[align="left"] i').textContent.trim().split(', '),
                    description: description_elements[description_elements.length - 1].textContent.trim(),
                });
            }

            return item_list;
        });

        if (page_items === null) {
            break;
        }

        author_items.concat(...page_items);

        if (!has_next_page) {
            break;
        }

        page_number++;
    }

    return author_items;
};


const get_author_refs = async (page, author) => {
    const author_refs = [];

    let page_number = 1;

    while (true) {
        await page.goto(`${AUTHOR_REFS_URL}?authorid=${author.id}&pagenum=${page_number}`);

        if ((await page.url()).endsWith('page_error.asp')) {
            throw new Error('Ошибка сервера, попробуйте повторить попытку');
        }

        const has_next_page = await page.evaluate(() => {
            const pagination_elements = document.querySelectorAll('#pages td.mouse-hovergr');

            if (pagination_elements.length === 0) {
                return false;
            }

            return pagination_elements[pagination_elements.length - 1].querySelector('font') === null;
        });

        const node_console = console;

        const page_refs = await page.evaluate(() => {
            const ref_row_list = document.querySelectorAll('#restab tr[valign="middle"]');
            const ref_list = [];

            if (ref_row_list.length === 0) {
                return null;
            }

            node_console.log(`Загрузка ссылок на статьи автора (страница ${page_number})...`);

            for (let row of ref_row_list) {
                ref_list.push({
                    id: row.id.slice(3),
                    count_number: row.querySelector('td[align="center"] b').textContent.trim(),
                    source: row.querySelector('td[align="left"] > font').textContent.trim(),
                    cite_item: row.querySelector('td[align="left"] table .menug').textContent.trim(),
                    malformed: row.querySelector('td[align="left"] > a') === null,
                });
            }

            return ref_list;
        });

        if (page_refs === null) {
            break;
        }

        author_refs.concat(...page_refs);

        if (!has_next_page) {
            break;
        }

        page_number++;
    }

    return author_refs;
};


const main = async () => {
    const browser = await puppeteer.launch({headless: false});
    const page = await browser.newPage();
    await page.goto('https://elibrary.ru/defaultx.asp');

    const cookies = JSON.parse(fs.readFileSync('cookies.json').toString());

    await page.setCookie(...cookies);

    if (!Object.keys(cookies).includes('SCookieID') && !Object.keys(cookies).includes('SUserID')) {
        const credentials = await prompt_login();

        try {
            await authenticate(page, credentials.login, credentials.password);
        } catch (exception) {
            console.log(exception.message);
            process.exit(1);
        }
    }

    const author = await get_author(page);

    if (author === null) {
        console.log('Автора с таким именем не найдено');
        process.exit(0);
    }

    const items = await get_author_items(page, author);
    let refs;

    try {
        refs = await get_author_refs(page, author);
    } catch (exception) {
        console.log(exception.message);
        process.exit(1);
    }

    let got_malformed_refs = false;

    for (let ref of refs) {
        if (ref.malformed) {
            got_malformed_refs = true;

            console.log('\nНайдена неверная ссылка на статью');
            console.log('Порядковый номер ссылки в списке ссылок: ' + ref.count_number);
            console.log('Название цитируемой статьи:\n' + ref.source);
            console.log('Название ссылающейся статьи:\n' + ref.cite_item);
        }
    }

    if (!got_malformed_refs) {
        console.log('Неверных ссылок не найдено');
    }
};


main().then();
